SignalMessaging/profiles/OWSUserProfile.m

```diff
@@ -0,0 +1,347 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "OWSUserProfile.h"
+#import "NSString+OWS.h"
+#import <SignalServiceKit/AppContext.h>
+#import <SignalServiceKit/Cryptography.h>
+#import <SignalServiceKit/NSData+hexString.h>
+#import <SignalServiceKit/NSNotificationCenter+OWS.h>
+#import <SignalServiceKit/TSAccountManager.h>
+#import <YapDatabase/YapDatabaseConnection.h>
+#import <YapDatabase/YapDatabaseTransaction.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+NSString *const kNSNotificationName_LocalProfileDidChange = @"kNSNotificationName_LocalProfileDidChange";
+NSString *const kNSNotificationName_OtherUsersProfileWillChange = @"kNSNotificationName_OtherUsersProfileWillChange";
+NSString *const kNSNotificationName_OtherUsersProfileDidChange = @"kNSNotificationName_OtherUsersProfileDidChange";
+
+NSString *const kNSNotificationKey_ProfileRecipientId = @"kNSNotificationKey_ProfileRecipientId";
+NSString *const kNSNotificationKey_ProfileGroupId = @"kNSNotificationKey_ProfileGroupId";
+
+NSString *const kLocalProfileUniqueId = @"kLocalProfileUniqueId";
+
+@interface OWSUserProfile ()
+
+@property (atomic, nullable) OWSAES256Key *profileKey;
+@property (atomic, nullable) NSString *profileName;
+@property (atomic, nullable) NSString *avatarUrlPath;
+@property (atomic, nullable) NSString *avatarFileName;
+@property (atomic, nullable) NSDate *lastUpdateDate;
+
+@end
+
+#pragma mark -
+
+@implementation OWSUserProfile
+
++ (OWSUserProfile *)getOrBuildUserProfileForRecipientId:(NSString *)recipientId
```

Comment:
We should define `+(NSString *)collection` to be the legacy class name so users don't have to re-build their profile key collection from scratch.

Reply:
Great point.